### ISSUE TITLE
Replace fog with featureful-sdk's fog

### ISSUE THREAD

**Issue by SaintWish** [MSRevive/MasterSwordRebirth#247]
I think it would be nice to replace our current fog with https://github.com/FreeSlave/halflife-featureful.

**Comment by frendship-loost** [comment]
https://twhl.info/wiki/page/Tutorial%3A_Coding_Fog

This fog also seems to work much better than our current setup. Consider it an option?

**Comment by frendship-loost** [comment]
reminder of the awful bug we currently suffer from
![image](https://github.com/user-attachments/assets/f67a8c7f-839a-4bfc-ba96-eb3881953fb9)

**Comment by tschumann** [comment]
Hm, unless I'm mistaken, the fog in https://github.com/FreeSlave/halflife-featureful/blob/86a78f0b388d528f1fa9d53a3937d63955ecfd8d/cl_dll/gl_dynamic.h#L8 and https://github.com/FreeSlave/halflife-featureful/blob/86a78f0b388d528f1fa9d53a3937d63955ecfd8d/cl_dll/tri.cpp#L97 looks like Half-Life's default fog with optional call-out to OpenGL's fog functions (which Half-Life uses under the hood anyway).

Same with The Whole Half-Life's fog tutorial - just the stock fog with Master Sword: Rebirth seems to use anyway.

**Comment by SaintWish** [comment]
Ah okay, so we should probably clean up the fog code then since HL uses the OpenGL under the hood anyways.

**Comment by tschumann** [comment]
Yeah although I don't think there's much that can change in the existing
implementation (unless more modern versions of OpenGL support more fog
parameters.

**Comment by tschumann** [comment]
Okay from a quick search I couldn't find a good reference but https://www.swiftless.com/tutorials/opengl/fog.html#google_vignette mentions glHint(GL_FOG_HINT, GL_NICEST); but other than that, using immediate mode OpenGL looks to be the same fog functions.

That said, the 25th anniversary Half-Life engine now ships with two shaders, both of which seem to do some things with fog. Is this currently running on the pre-25th anniversary engine?

**Comment by SaintWish** [comment]
Yeah currently we're running on the pre HL25 engine. Though it shouldn't be too difficult to switch we just never planned on it due to time constraints and no changes to the limits we are close to.

We might switch in the future though, though that being said I did cherry pick some fixes from the SDK to ours.

**Comment by tschumann** [comment]
Okay - might be worth just trying it with the new engine and seeing if it works? I think there were a few early breaks but some updates to make sure that mods worked afterwards.

**Comment by SaintWish** [comment]
Yeah might be worth trying out, the only thing I can foresee being an issue is the ReHLDS fork were using breaking.

**Comment by frendship-loost** [comment]
Our fog codebase is very confusing and large pieces of it (the time of day system) are almost entirely obsolete. Maybe re-implement from the ground up?

The TWHL tutorial example has fog blending working pre and post HL-25th update. AND it correctly shades transparent triangles somehow... my understanding is low here.

```
void CFog::HUD_DrawTransparentTriangles( void )
{
	// At this stage we want to set black fog, so beams are faded out
	SetGLFog(Vector(0, 0, 0));
}

void CFog::BlendFog( void )
{
    if(!m_fogChangeTime && !m_fogBlendTime)
        return;

    // Clear if blend is over
    if((m_fogChangeTime+m_fogBlendTime) < gEngfuncs.GetClientTime())
    {
        memcpy(&m_fogParams, &m_fogBlend2, sizeof(fog_params_t));
        memset(&m_fogBlend1, 0, sizeof(fog_params_t));
        memset(&m_fogBlend2, 0, sizeof(fog_params_t));

        m_fogChangeTime = 0;
        m_fogBlendTime = 0;
        return;
    }

    // Perform linear blending
    float interp = (gEngfuncs.GetClientTime() - m_fogChangeTime)/m_fogBlendTime;
    VectorScale(m_fogBlend1.color, (1.0 - interp), m_fogParams.color);
    VectorMA(m_fogParams.color, interp, m_fogBlend2.color, m_fogParams.color);
    m_fogParams.enddist = (m_fogBlend1.enddist * (1.0 - interp)) + (m_fogBlend2.enddist * interp);
    m_fogParams.startdist = (m_fogBlend1.startdist * (1.0 - interp)) + (m_fogBlend2.startdist * interp);
} 
```

I believe this is the piece that performs blending on transparent triangles in the TWHL fog tutorial in fog.cpp? But I need some assistance here. Our fog code does nothing with transparent triangles as far as I can read, it only fogs solid triangles, underwater, and in mirrors.


Another example of the bug affecting transparent triangles pipeline:
![Image](https://github.com/user-attachments/assets/f4972e13-661b-41f4-8204-85d246534358)


Working in TWHL example (see the transparent gate):
![Image](https://github.com/user-attachments/assets/ef825b4b-f44a-46f1-bac1-f5d0428138cd)

**Comment by tschumann** [comment]
What is a map with fog in it? It looks like the fog code is being called twice maybe?

**Comment by frendship-loost** [comment]
You can call fog at anytime with "setweather snow" in console, or setting it to any valid weather.

**Comment by tschumann** [comment]
Thanks - that worked. So it looks like the call to `RenderFog` in `V_CalcRefdef` renders fog but the call in `HUD_DrawTransparentTriangles` does nothing and I think that's the possible cause of this issue. In normal Half-Life, the call should be in `HUD_DrawTransparentTriangles` (which I guess makes sure that it renders after models), but I haven't fully figured out the custom renderer.

I looked at replacing the OpenGL calls with this:
```
gEngfuncs.pTriAPI->Fog(CEnvMgr::m_Fog.Color, CEnvMgr::m_Fog.Start, CEnvMgr::m_Fog.End, CEnvMgr::m_Fog.Enabled);
gEngfuncs.pTriAPI->FogParams(CEnvMgr::m_Fog.Density, 1);
```
But the result was black fog instead of white fog. Something to figure out anyway as using the engine's fog will probably be best when moving the 25th anniversary engine so it can work with the shader fog.

**Comment by SaintWish** [comment]
Yeah once we make the move to 25th update we can probably get rid of the custom renderer stuff.

Before doing that though there's a few important things I wanna get done

**Comment by SaintWish** [comment]
Closed in favor of https://github.com/MSRevive/MasterSwordRebirth/issues/289
